update_test_checks: add new --filter-out-after option

david-arm opened the pull request:
Whilst trying to clean up some loop vectoriser IR tests (see
test/Transforms/LoopVectorize/AArch64/partial-reduce-chained.ll
for example) a reviewer on PR #129047 suggested it would be
nice to have an option to stop generating CHECK lines after a
certain point. Typically when performing a transformation with
the loop vectoriser we don't usually care about any CHECK lines
generated for the scalar tail of the loop, since the scalar
loop is kept intact. Previously if you wanted to eliminate such
unwanted CHECK lines you had to run the update script, then
manually delete all the lines corresponding to the scalar loop.
This can be very time consuming if the tests ever need changing.

What I've tried to do here is add a new --filter-out-after
option alongside the existing --filter* options that provides
support for stopping the generation of any CHECK lines beyond
the line that matches the filter. With the existing filter
options we never generate CHECK-NEXT lines, but we still care
about ordering with --filter-out-after so I've amended the
code to ensure we treat this filter differently.

fhahn commented on llvm/test/tools/UpdateTestChecks/update_test_checks/Inputs/filter-all-after.ll:
```diff
@@ -0,0 +1,17 @@
+; RUN: opt < %s -S | FileCheck %s
+
+define i32 @func({i32, i32} %x, i32 %y, i1 %cond) {
```
Could you add a test with another function, to check that the filter applies per-function only?

fhahn commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -151,6 +159,13 @@ def __call__(self, parser, namespace, values, option_string=None):
         help="Exclude lines matching REGEX",
     )
+    filter_group.add_argument(
+        "--filter-all-after",
```
The existing `--filter` is used to specify a filter to include lines matching it and `--filter-out` to exclude. Something like `--filter-out-after` may be more accurate?

fhahn commented on llvm/utils/UpdateTestChecks/common.py:
```diff
@@ -151,6 +159,13 @@ def __call__(self, parser, namespace, values, option_string=None):
+        action=FilterAction,
+        dest="filters",
+        metavar="REGEX",
+        help="Exclude all lines after line matching REGEX",
```
This should clarify that this is per-function?

hnrklssn commented:
How does this work for `update_cc_test_checks`? Does the regex match the lines in the source input, or in the LLVM IR output?

david-arm commented:
> How does this work for `update_cc_test_checks`? Does the regex match the lines in the source input, or in the LLVM IR output?

Hmm, I'm not sure. I'll have a look. If it doesn't make sense for source code, is it ok to have a update_test_checks specific flag?

hnrklssn commented:
> > How does this work for `update_cc_test_checks`? Does the regex match the lines in the source input, or in the LLVM IR output?
> 
> Hmm, I'm not sure. I'll have a look. If it doesn't make sense for source code, is it ok to have a update_test_checks specific flag?

I'd be fine with limiting it to update_test_checks, just wanted to make sure that it works in the instances where it's available. Of course, if we can get it in the other scripts for free that's great, too! That said, for update_test_checks, does the regex match the input or output? Maybe it would be good with a test case where the input has anonymous instructions, and the output is after running `instnamer`? Then matching against something that is only in the input but not the output, or vice versa.

david-arm committed:
update_test_checks: add new --filter-all-after option

Whilst trying to clean up some loop vectoriser IR tests (see
test/Transforms/LoopVectorize/AArch64/partial-reduce-chained.ll
for example) a reviewer on PR #129047 suggested it would be
nice to have an option to stop generating CHECK lines after a
certain point. Typically when performing a transformation with
the loop vectoriser we don't usually care about any CHECK lines
generated for the scalar tail of the loop, since the scalar
loop is kept intact. Previously if you wanted to eliminate such
unwanted CHECK lines you had to run the update script, then
manually delete all the lines corresponding to the scalar loop.
This can be very time consuming if the tests ever need changing.

What I've tried to do here is add a new --filter-out-after
option alongside the existing --filter* options that provides
support for stopping the generation of any CHECK lines beyond
the line that matches the filter. With the existing filter
options we never generate CHECK-NEXT lines, but we still care
about ordering with --filter-out-after so I've amended the
code to ensure we treat this filter differently.

david-arm commented:
> I'd be fine with limiting it to update_test_checks, just wanted to make sure that it works in the instances where it's available. Of course, if we can get it in the other scripts for free that's great, too! That said, for update_test_checks, does the regex match the input or output? Maybe it would be good with a test case where the input has anonymous instructions, and the output is after running `instnamer`? Then matching against something that is only in the input but not the output, or vice versa.

Thanks for the suggestions! I've added a test for update_cc_test_checks to show that the filtering happens only on the output when checking C files. I've also added a filter-out-after3.ll test for update_test_checks using your instname suggestion, which demonstrates that filtering does not work on the input IR.

hnrklssn reviewed (approved):
LGTM
